lessons/python_packages/lesson.md

```diff
@@ -0,0 +1,258 @@
+     |__ __init__.py
+```
+
+But at the moment, we've just got some flat files.
+
+```
+project
+|
+|__ norms.py
+|__ metrics.py
+```
+
+So, the first step is to move files around. First comes the hardest part: choosing a package name. I'll call mine `measure`. Create a directory with that name, and move the python files in there.
+
+```
+project
+|
+|__ measure
+     |__ norms.py
+     |__ metrics.py
+```
+
+There is one more crucial file: `__init__.py` lets the Python interpreter know that there are importable modules in this directory. This is the script that gets run when you execute `import measure`. For more about what you can do with modules, you can see the (Python docs)[https://docs.python.org/3/tutorial/modules.html]. After adding `__init__.py`, the project directory should be
+
+```
+project
+|
+|__ measure
+     |__ __init__.py
+     |__ norms.py
+     |__ metrics.py
+```
+
+
+## 2. setup.py
+
+At this point, the library can be imported if we're in the same directory, but it isn't a package. To let setuptools and `pip` know how to handle it, we need to add the `setup.py` file.
+
+A very basic version of `setup.py` is
+
+```
+from setuptools import setup
+
+setup(
+    # Whatever arguments you need/want
+)
+```
+
+If you were to run that file, you'd get a whole bunch of warnings, and nothing would actually get packaged. As a bare minimum, here is our `setup.py` for it to work.
+
+```
+from setuptools import setup
+
+setup(
+    # Needed to silence warnings
+    name='Measurements',
+    url='https://github.com/jladan/package_demo',
+    author='John Ladan',
+    author_email='jladan@uwaterloo.ca',
+    # Needed to actually package something
+    packages=['measure'],
+    # Needed for dependencies
+    install_requires=['numpy'],
+    # *strongly* suggested for sharing
+    version='0.1',
+    license='MIT',
+    description='An example of a python package from pre-existing code',
+    # We will also need a readme eventually (there will be a warning)
+    # long_description=open('README.txt').read(),
+)
+```
+
+## 3. Installing?
+
+Technically, you can install the module by running `python setup.py install`, however it's much more convenient to wrap it all up and distribute online.
+
+### Using git
+
+This is my preferred way of sharing/storing/installing packages. Simply create a git repo of the project directory, and put it somewhere. Then, use `pip` to install it from that repo directly
+
+```
+pip install git+https://github.com/jladan/package_demo.git#egg=measurements
+```
+
+An advantage of this is that you can also install to particular branches, tags, and commits. A common practice is add a tag for the version release `git tag v0.1`. Users of the library can then lock into that particular version,
+
+```
+pip install git+https://github.com/jladan/package_demo.git@v0.1#egg=measurements
+```
+
+
+### Distributing on PyPI
+
+In my opinion, distributing on PyPI is more complicated than with git, and not much more useful.
```

Comment:
It's a supportive statement: that's why it is my preferred method.

I felt the technical intricacies of how `pip` works were beyond the scope of this lesson.